### DISCUSSION TITLE
Generating mwm files with custom data

### DISCUSSION THREAD

**Discussion by cbosdo** [organicmaps/organicmaps#6361] [Q&A]
Hello,

I have recently discovered organic maps to your Google Summer of Code representative at the mentor summit and started using it. I have a piece of code [generating osmand data](https://github.com/cbosdo/osmand-fuel-price) file from the fuel prices published by the french government and would like to generate those data for Organic Maps. I have found the `maps_generator` tool, but I still fail to see how to go from `.osm` files to `.mwm`. Would you have some code pointers to get me started in that task?

My existing code uses osmand's tools, which are storing everything in memory. Ideally I would like to process the data in a stream to run it on my RPi B... but that would be the cherry on the cake.

**Comment by boswellp** [comment]
I managed to create .mwm files (from .pbf downloads from Geofabrik) on Ubuntu 18.04. I'll write up the process. My problem now is to somehow store an .mwm file in the organicmaps  app data folder. One can get at the data folder by connecting an Android phone to a PC (I use a Pixel 4a phone).

**Reply by cbosdo** [reply]
That would be awesome for newcomers like me, thanks a lot.

**Reply by twmaps** [reply]
Hi @boswellp can you share the guidelines or documentation or creating .mwm files from .pbf downloads from Geofabrik? I am try to develop route able app for my university project.

**Reply by AhmadWork** [reply]
Hi @boswellp
I'm working on a project that try to add extra data for the mwm maps and would love to know the process you used to generate the mwm files.

**Reply by manylittlestorms** [reply]
> I managed to create .mwm files (from .pbf downloads from Geofabrik) on Ubuntu 18.04. I'll write up the process. My problem now is to somehow store an .mwm file in the organicmaps  app data folder. One can get at the data folder by connecting an Android phone to a PC (I use a Pixel 4a phone).

Can you share a step by step guide? I am interested too.

**Reply by aycmaen48** [reply]
@boswellp Have you managed to write up the process of converting .pbf to .mwm by any chance? I'd be curious.

**Comment by cbosdo** [comment]
The more I look into the code and the app, the more it seems this could be a cool feature: getting gas prices in the `PlacePageView` could be nice. I think I could somehow try to match the data with the prices with the OSM nodes to provide those additional data directly in the existing gas station objects. I have no clue how extensible is this would be possible / desired to include these data in the `mwm` files and the UI. It's clearly not a simple project sadly.

Another problem is the data accuracy from french gas prices... some stations have wrong lat/lng and searching may be tricky when there is a station on each side of a road. In the Osmand version I was just creating a new gas station object with the prices as name. May be that would be a first step here?

**Reply by biodranik** [reply]
Is it possible to properly link gas prices to OSM gas station objects? It would be great to have a service that independently pulls/renews gas prices and that points to OSM objects. And have a validator to keep it in sync.

**Reply by cbosdo** [reply]
I just discovered that quite a lot of the OSM gas station objects in france have a `ref:FR:prix-carburants` tag with the ID of the station on the price data. Providing a service mapping those would be possible, though there are still 255 on the 9702 gas stations listed on prix-carburants.gouv.fr that have no match in OSM objects. I'll try to loop through them and report them where missing.

What would you imagine such a service to take as input and provide as a return? I wonder if there are other open data to be used for other countries...

**Reply by cbosdo** [reply]
I discovered https://openfuelmap.net while investigating. It seems broken, but gets data for Germany too.

**Reply by biodranik** [reply]
A separate service can be implemented that periodically queries different countries' data, with a simple API to retrieve the price based on some reference in OSM.

**Reply by cbosdo** [reply]
As long as we could get those data for offline use that's fine for me.

I also found this week that Portugal may have data we could use: https://precoscombustiveis.dgeg.gov.pt

**Reply by cbosdo** [reply]
Here is a first attempt at a service implementing such an API: feel free to comment on it: https://github.com/cbosdo/gas-price. Note that I am not hosting it anywhere for now.

**Reply by cbosdo** [reply]
The fetcher reports the inconsistencies between prix-carburants.gouv.fr and OSM. I started addressing some of them, but there are about 280 of them

**Reply by wadouk** [reply]
Hi, I manage https://openfuelmap.net
Get some trouble but all was fixed.
Which kind of inconsistencies did you find ?

**Reply by cbosdo** [reply]
> Hi, I manage https://openfuelmap.net Get some trouble but all was fixed.

Nice ! I'll take a look. I wonder if we could somehow merge my approach and yours.

> Which kind of inconsistencies did you find ?

You can build and run the fr fetcher I wrote: the output reports gas stations that have a `ref:FR:prix-carburants` tag not matching an ID on prix-carburants anymore and also reports those from prix-carburant that can't match an OSM node or way with `amenity:fuel` tag. I have some changes locally that try to match for a nearby node to suggest changes in OSM data.

My goal was to have something as lightweight as possible: that's why I didn't go with a GIS postgres like what you did. What we need for this discussion is an API that provides the prices from an OSM node/way ID... this it probably something that could be done with openfuelmap. I wonder however how we could use those data offline in organicmaps.

**Reply by wadouk** [reply]
I can make available to download the full data merged / conflation between both sources.
I have [an issue](https://framagit.org/PanierAvide/OpenFuelMap/-/issues/23) to match every day gas stations.

Maybe some osmose analyse could help to do the trick. [Already one exist](https://github.com/osm-fr/osmose-backend/blob/040bf3d03da7b74be5396162fe8d70982abbb821/analysers/analyser_merge_fuel_FR.py#L32C48-L32C48) to map gas stations.

Having an API endpoint to get data for an osm id could be done quiet easliy. [PR welcome](https://framagit.org/PanierAvide/OpenFuelMap).

**Comment by cbosdo** [comment]
Note that I managed to work around my original problem by exporting the french gas prices as KML files. It's quite nice to use, but still requires to manually import the files
